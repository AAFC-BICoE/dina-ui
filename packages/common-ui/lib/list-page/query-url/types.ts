export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
