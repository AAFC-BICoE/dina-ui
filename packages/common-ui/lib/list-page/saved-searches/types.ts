import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
