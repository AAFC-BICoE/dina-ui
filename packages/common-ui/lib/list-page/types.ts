import { KitsuResource } from "kitsu";
import { Column } from "react-table";

/**
 * This type extends the react-table column type, this just adds a few specific fields for elastic
 * search mapping and internationalization.
 */
export interface TableColumn<TData extends KitsuResource>
  extends Column<TData> {
  /**
   * User-friendly column to be displayed. You can use a DinaMessage key for internationalization.
   */
  label?: string;

  /**
   * Elastic search path to the attribute.
   *
   * Example: `data.attributes.name`
   */
  attributePath?: string;

  /**
   * This field is used to find the relationship in the included section.
   */
  relationshipType?: string;

  /**
   * Is this attribute considered a keyword in elastic search. Required for filtering and sorting.
   */
  isKeyword?: boolean;

  /**
   * The QueryPage will only display the accessors that are displayed on the result table. However,
   * if you have custom cells that receive other fields you will need to add them to this list so
   * elastic search includes the fields in the result.
   *
   * Example: `data.attributes.name`
   *
   * Please note that duplicate fields are automatically removed so you don't need to worry about
   * having unique accessors.
   */
  additionalAccessors?: string[];
}

/**
 * The full path will be generated for elastic using a combination of the parent path and
 * the value. The path is generated using the following:
 *
 * {parentPath}.{path}.{value}
 *
 * Example: included.attributes.determination.verbatimScientificName
 */
export interface ESIndexMapping {
  /**
   * Name of the attribute.
   *
   * Example: verbatimScientificName
   */
  value: string;

  /**
   * Text that is displayed to the user in the Query Filtering option menu.
   *
   * This text is a user-friendly generated label, which may show some paths to help the user
   * understand the relationships better. This is generated from the path.
   *
   * Example: determination.verbatimScientificName
   */
  label: string;

  /**
   * The attributes type. This can change how the query row is displayed and the options provided.
   *
   * Examples: text, keyword, boolean, date, boolean, long, short, integer...
   */
  type: string;

  /**
   * If enabled, it will allow the user to see suggestions as they type. The suggestions will come
   * from elastic search based on most common values saved.
   *
   * Only available for the text type.
   */
  distinctTerm: boolean;

  /**
   * If enabled it will allow the user to search based on the starting of a word.
   *
   * Example: Hexapoda can be matched with "hex".
   */
  startsWithSupport: boolean;

  /**
   * If enabled it will allow the user to search based in the middle of a word.
   *
   * Example: Hexapoda can be matched with "pod".
   */
  containsSupport: boolean;

  /**
   * If enabled it will allow the user to search based on the ending of a word.
   *
   * Example: Hexapoda can be matched with "poda".
   */
  endsWithSupport: boolean;

  /**
   * The path for the attribute without the attribute name. This path does not include the parent
   * path.
   *
   * Example: attribute.determination
   */
  path: string;

  /**
   * If the attribute belongs to a relationship, this is the path for only the parent. When generating
   * the elastic search query it will use this as the prefix of the path.
   *
   * Example: included
   */
  parentPath?: string;

  /**
   * If the attribute belongs to a relationship, this is the name which will be used to group
   * attributes under the same relationship together in the search. This name will also be used to
   * display text of the group.
   *
   * This text should match the user-friendly label in the locales files. It will be searched using
   * title_[parentName].
   *
   * Example: preparationMethod (and the label will be `title_preparationMethod`)
   */
  parentName?: string;

  /**
   * The parent type is the relationship type to be used. This will be used for elastic search
   * filtering. The reason this is used since the same field can be used for multiple relationships.
   *
   * Example: preparation-method
   */
  parentType?: string;

  /**
   * Only provided if it was added using a dynamic field config.
   */
  dynamicField?: DynamicField;
}

/**
 * Each type on the query builder has a function that is used to transform the query builder row
 * into elastic search logic.
 *
 * @see QueryBuilderElasticSearchExport
 * @see QueryRowTextSearch
 */
export interface TransformToDSLProps {
  /**
   * The query builder type, not really used but required in the function for the QueryBuilder
   * library.
   *
   * For the Elastic Search type you can use the fieldInfo.type instead.
   */
  queryType: string;

  /**
   * The value to search against. If the type is "text" then this is the text that will be filtered.
   */
  value: string;

  /**
   * The operation being performed.
   *
   * For example:
   * "equals", "notEquals", "containsText"
   *
   * Operators are defined in the QueryBuilderConfig file.
   */
  operation: string;

  /**
   * This field path is the unique path given to each item in the field list.
   *
   * For getting the proper absolute field path, use the FieldInfo to generate it.
   */
  fieldPath: string;

  /**
   * The elastic search mapping for the field.
   */
  fieldInfo?: ESIndexMapping;
}

export type DynamicFieldType = "managedAttribute" | "fieldExtension";

export interface DynamicFieldsMappingConfig {
  /** Attribute level dynamic fields */
  fields: DynamicField[];

  /** Dynamic fields for relationships */
  relationshipFields: RelationshipDynamicField[];
}

export interface DynamicField {
  /**
   * Option label that should be used.
   */
  label: string;

  type: DynamicFieldType;

  path: string;

  /**
   * Endpoint where these dynamic fields can be retrieved to list.
   *
   * Example: "collection-api/managed-attribute"
   */
  apiEndpoint: string;

  /**
   * Optional field to indicate which Managed Attributes or Field Extensions should be listed.
   */
  component?: string;
}

/**
 * Configuration for where the Dynamic Field can be found within the relationship index mapping.
 */
export interface RelationshipDynamicField extends DynamicField {
  referencedBy: string;
  referencedType: string;
}
