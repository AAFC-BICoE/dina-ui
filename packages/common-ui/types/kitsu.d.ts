/// <reference types="node" />

import { AxiosInstance } from "axios";
import { SetRequired } from "type-fest";

declare module "kitsu" {
  // export default Kitsu;

  /** JSONAPI client. */
  export default class Kitsu {
    public axios: AxiosInstance;

    public headers: any;

    constructor(params: KitsuConstructorParams);

    public delete(...args: any[]): Promise<any>;

    public get<TData, TMeta = undefined>(
      path: string,
      params: GetParams
    ): Promise<KitsuResponse<TData, TMeta>>;

    public patch(...args: any[]): Promise<any>;

    public post(...args: any[]): Promise<any>;
  }

  /** Params for the Kitsu JSONAPI client's constructor */
  export interface KitsuConstructorParams {
    baseURL: string;
    headers?: any;
    pluralize?: boolean;
    camelCaseTypes?: boolean;
    resourceCase?: "kebab" | "snake" | "none";
  }

  /** Parameters for GET requests. */
  export interface GetParams {
    /** Fields to include in the response data. */
    fields?: FieldsParam;

    /** Resource filter */
    filter?: FilterParam;

    /**
     * Sort order + attribute.
     * Examples:
     *  - name
     *  - -description
     */
    sort?: string;

    /** Included resources. */
    include?: string;

    /** Vendor-specific parameter for paginating listed data. */
    page?: any;
  }

  /** Parameter for requesting sparse fields. */
  export interface FieldsParam {
    [key: string]: string;
  }

  /** Parameter for filtering listed data. */
  export type FilterParam = string | Record<string, string>;

  /** The response from a Kitsu GET request. */
  export interface KitsuResponse<
    TData extends KitsuResponseData,
    TMeta = undefined
  > {
    data: TData extends (infer R)[]
      ? PersistedResource<R>[]
      : PersistedResource<TData>;
    meta: TMeta;
  }

  /** The Kitsu response data can either be one resource or an array of resources. */
  export type KitsuResponseData = KitsuResource | KitsuResource[];

  /** JSONAPI resource base attributes. */
  export interface KitsuResource {
    id?: string;
    type: string;
  }

  export interface KitsuResourceLink {
    id: string | null;
    type: string;
  }

  /**
   * Makes the 'id' field required on a resource type and all of its relationships.
   * Used when assuming that data from the back-end always has the ID set.
   */
  export type PersistedResource<TData extends KitsuResource = KitsuResource> = {
    [P in keyof TData]: TData[P] extends KitsuResource
      ? PersistedResource<TData[P]>
      : TData[P] extends KitsuResource | undefined
      ? PersistedResource<TData[P]> | undefined
      : TData[P];
  } &
    Required<KitsuResource>;

  /**
   * Used when creating or updating a resource.
   * Makes the 'id' field optional on the main resource.
   * Makes the 'id' field required on linked resources.
   * 'type' must be defined.
   */
  export type InputResource<TData extends KitsuResource> = SetRequired<
    {
      [P in keyof TData]?: TData[P] extends KitsuResource | undefined
        ? Required<KitsuResourceLink> | undefined
        : TData[P];
    },
    "type"
  >;
}
