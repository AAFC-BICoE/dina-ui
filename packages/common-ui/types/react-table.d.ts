import "@tanstack/react-table";

declare module "@tanstack/table-core" {
  interface ColumnMeta<TData extends RowData, TValue> {
    className?: string;
    style?: CSSProperties;
  }

  interface TableMeta<TData extends RowData> {
    updateData?: (rowIndex: number, columnId: string, value: unknown) => void;
    enableEditing?: boolean;
  }
}
