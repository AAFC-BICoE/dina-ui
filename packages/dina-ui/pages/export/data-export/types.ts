export interface SavedExportColumnStructure {
  name: string;
  component: string;
  columns: string[];
}
