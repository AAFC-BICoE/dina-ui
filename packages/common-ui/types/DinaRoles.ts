export const DINA_ADMIN = "dina-admin";
export const READ_ONLY_ADMIN = "read-only-admin";
export const SUPER_USER = "super-user";
export const USER = "user";
export const GUEST = "guest";
export const READ_ONLY = "read-only";
