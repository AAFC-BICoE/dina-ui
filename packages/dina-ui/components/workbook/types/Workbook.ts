/**
 * A specific row on a spreadsheet.
 */
export interface WorkbookRow {
  rowNumber: number;
  content: string[];
}

/**
 * A spreadsheet can contain multiple sheets, each sheet has an array of rows.
 */
export interface WorkbookJSON {
  [sheetNumber: number]: WorkbookRow[];
}
