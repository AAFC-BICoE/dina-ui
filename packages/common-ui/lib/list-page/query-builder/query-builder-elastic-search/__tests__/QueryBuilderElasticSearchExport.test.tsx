import { KitsuResource } from "kitsu";
import { TableColumn } from "../../../types";
import {
  applyGroupFilters,
  applyPagination,
  applyRootQuery,
  applySortingRules,
  applySourceFiltering,
  includedTypeQuery,
  termQuery,
  matchQuery,
  existsQuery,
  rangeQuery,
  prefixQuery,
  infixQuery,
  suffixQuery
} from "../QueryBuilderElasticSearchExport";

const ELASTIC_SEARCH_QUERY: any = {
  query: {
    bool: {
      must: [
        {
          term: {
            "data.attributes.materialSampleName.keyword": "Test"
          }
        },
        {
          term: {
            "data.attributes.preparationDate": "2022-11-30"
          }
        }
      ]
    }
  }
};

const columnDefinitions: (TableColumn<KitsuResource> | string)[] = [
  {
    label: "testColumn1",
    accessor: "data.attributes.testColumn1",
    isKeyword: true,
    additionalAccessors: [
      "data.attributes.includeThis1",
      "data.attributes.includeThis2"
    ]
  },
  {
    label: "testColumn2",
    accessor: "data.attributes.testColumn2",
    isKeyword: false,
    additionalAccessors: ["data.attributes.includeThis3"]
  },
  "testColumn3",
  {
    label: "testColumn4",
    accessor: "data.attributes.testColumn4",
    isKeyword: false,
    relationshipType: "relationshipType1"
  },
  {
    label: "testColumn5",
    accessor: "data.attributes.testColumn5",
    isKeyword: true,
    relationshipType: "relationshipType2"
  }
];

describe("QueryBuilderElasticSearchExport functionality", () => {
  describe("applyPagination", () => {
    test("Pagination is correctly applied to query", async () => {
      // Pagination should be added to the existing query without altering anything.
      expect(applyPagination(ELASTIC_SEARCH_QUERY, 25, 0)).toMatchSnapshot();
      expect(applyPagination(ELASTIC_SEARCH_QUERY, 100, 100)).toMatchSnapshot();
    });
  });

  describe("applySortingRules", () => {
    test("Basic sorting is correctly applied to the query", async () => {
      // Single Sorting rule, descending.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [{ id: "data.attributes.testColumn1", desc: true }],
          columnDefinitions as any
        )
      ).toMatchSnapshot();

      // Single Sorting rule, ascending.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [{ id: "data.attributes.testColumn1", desc: false }],
          columnDefinitions as any
        )
      ).toMatchSnapshot();

      // Multiple sorting rules.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [
            { id: "data.attributes.testColumn1", desc: false },
            { id: "data.attributes.testColumn2", desc: true }
          ],
          columnDefinitions as any
        )
      ).toMatchSnapshot();
    });

    test("Relationship sorting query generation", async () => {
      // Single Relationship Sorting rule, descending.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [{ id: "data.attributes.testColumn4", desc: true }],
          columnDefinitions as any
        )
      ).toMatchSnapshot();

      // Multiple Relationship Sorting rules.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [
            { id: "data.attributes.testColumn4", desc: true },
            { id: "data.attributes.testColumn5", desc: false }
          ],
          columnDefinitions as any
        )
      ).toMatchSnapshot();

      // Finally, a mix of normal and relationship sorting rules.
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [
            { id: "data.attributes.testColumn1", desc: false },
            { id: "data.attributes.testColumn2", desc: true },
            { id: "data.attributes.testColumn4", desc: true },
            { id: "data.attributes.testColumn5", desc: false }
          ],
          columnDefinitions as any
        )
      ).toMatchSnapshot();
    });

    test("No sorting to be performed, leave query the same without adding anything.", async () => {
      // Single Sorting rule, descending.
      expect(
        applySortingRules(ELASTIC_SEARCH_QUERY, [], columnDefinitions as any)
      ).toMatchSnapshot();
    });

    test("Attempting to sort on a column that does not exist", async () => {
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [{ id: "data.attributes.unknownColumn", desc: false }],
          columnDefinitions as any
        )
      ).toMatchSnapshot();
    });

    test("Strings as column definitions", async () => {
      expect(
        applySortingRules(
          ELASTIC_SEARCH_QUERY,
          [{ id: "testColumn3", desc: false }],
          columnDefinitions as any
        )
      ).toMatchSnapshot();
    });
  });

  describe("applySourceFiltering", () => {
    test("Ensure all source filtering are being applied to the query", async () => {
      expect(
        applySourceFiltering(ELASTIC_SEARCH_QUERY, columnDefinitions as any)
      ).toMatchSnapshot();
    });
  });

  describe("applyGroupFilters", () => {
    test("Single group", async () => {
      expect(
        applyGroupFilters(ELASTIC_SEARCH_QUERY, ["aafc"])
      ).toMatchSnapshot();
    });

    test("Multiple groups", async () => {
      expect(
        applyGroupFilters(ELASTIC_SEARCH_QUERY, ["aafc", "cnc", "seqdb"])
      ).toMatchSnapshot();
    });

    test("No groups", async () => {
      expect(applyGroupFilters(ELASTIC_SEARCH_QUERY, [])).toMatchSnapshot();
    });
  });

  describe("applyRootQuery", () => {
    test("Boolean logic exists, move it properly to it's own boolean section.", async () => {
      expect(
        applyRootQuery({
          bool: {
            must: [
              {
                term: {
                  "data.attributes.materialSampleName.keyword": "Test"
                }
              },
              {
                term: {
                  "data.attributes.preparationDate": "2022-11-30"
                }
              }
            ]
          }
        })
      ).toMatchSnapshot();
    });

    test("Boolean logic exists, also contains should logic, include the must match minimum.", async () => {
      expect(
        applyRootQuery({
          bool: {
            should: [
              {
                term: {
                  "data.attributes.materialSampleName.keyword": "Test"
                }
              },
              {
                term: {
                  "data.attributes.preparationDate": "2022-11-30"
                }
              }
            ],
            must: [
              {
                term: {
                  "data.attributes.materialSampleName.keyword": "Test"
                }
              },
              {
                term: {
                  "data.attributes.preparationDate": "2022-11-30"
                }
              }
            ]
          }
        })
      ).toMatchSnapshot();
    });

    test("No boolean logic exists, just return the query as is.", async () => {
      expect(applyRootQuery({ query: {} })).toMatchSnapshot();
    });
  });

  describe("Query helper functions", () => {
    test("includedTypeQuery", async () => {
      expect(includedTypeQuery("parentTest")).toMatchSnapshot();
    });

    test("termQuery", async () => {
      expect(termQuery("fieldTest", "valueToMatch", true)).toMatchSnapshot();
      expect(termQuery("fieldTest", "valueToMatch", false)).toMatchSnapshot();
    });

    test("matchQuery", async () => {
      expect(matchQuery("fieldTest", "valueToMatch")).toMatchSnapshot();
    });

    test("existsQuery", async () => {
      expect(existsQuery("fieldTest")).toMatchSnapshot();
    });

    test("rangeQuery", async () => {
      expect(rangeQuery("fieldTest", { lt: 500 })).toMatchSnapshot();
    });
  });

  describe("Partial matching query helper functions", () => {
    test("prefixQuery attribute", async () => {
      expect(
        prefixQuery(
          "data.attribute.materialSampleName",
          "searchValue",
          undefined
        )
      ).toMatchSnapshot();
    });

    test("prefixQuery relationship", async () => {
      expect(
        prefixQuery(
          "included.attributes.dwcRecordNumber",
          "searchValue",
          "collecting-event"
        )
      ).toMatchSnapshot();
    });

    test("infixQuery attribute", async () => {
      expect(
        infixQuery(
          "data.attribute.materialSampleName",
          "searchValue",
          undefined
        )
      ).toMatchSnapshot();
    });

    test("infixQuery relationship", async () => {
      expect(
        infixQuery(
          "included.attributes.dwcRecordNumber",
          "searchValue",
          "collecting-event"
        )
      ).toMatchSnapshot();
    });

    test("suffixQuery attribute", async () => {
      expect(
        suffixQuery(
          "data.attribute.materialSampleName",
          "searchValue",
          undefined
        )
      ).toMatchSnapshot();
    });

    test("suffixQuery relationship", async () => {
      expect(
        suffixQuery(
          "included.attributes.dwcRecordNumber",
          "searchValue",
          "collecting-event"
        )
      ).toMatchSnapshot();
    });

    test("Empty values are left as empty queries", async () => {
      expect(
        prefixQuery("data.attribute.materialSampleName", "", undefined)
      ).toStrictEqual({});
      expect(
        infixQuery("data.attribute.materialSampleName", "", undefined)
      ).toStrictEqual({});
      expect(
        suffixQuery("data.attribute.materialSampleName", "", undefined)
      ).toStrictEqual({});
    });
  });
});
