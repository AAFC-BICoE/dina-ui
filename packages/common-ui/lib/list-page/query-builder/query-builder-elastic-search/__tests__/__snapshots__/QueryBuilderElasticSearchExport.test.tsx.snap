// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions infixQuery attribute 1`] = `
Object {
  "match": Object {
    "data.attribute.materialSampleName.infix": Object {
      "query": "searchValue",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions infixQuery relationship 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "match": Object {
              "included.attributes.dwcRecordNumber.infix": Object {
                "query": "searchValue",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collecting-event",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions prefixQuery attribute 1`] = `
Object {
  "prefix": Object {
    "data.attribute.materialSampleName.prefix": "searchvalue",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions prefixQuery relationship 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.dwcRecordNumber.prefix": "searchvalue",
            },
          },
          Object {
            "term": Object {
              "included.type": "collecting-event",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions suffixQuery attribute 1`] = `
Object {
  "prefix": Object {
    "data.attribute.materialSampleName.prefix_reverse": "eulavhcraes",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Partial matching query helper functions suffixQuery relationship 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.dwcRecordNumber.prefix_reverse": "eulavhcraes",
            },
          },
          Object {
            "term": Object {
              "included.type": "collecting-event",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions existsQuery 1`] = `
Object {
  "exists": Object {
    "field": "fieldTest",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions includedTypeQuery 1`] = `
Object {
  "term": Object {
    "included.type": "parentTest",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions matchQuery 1`] = `
Object {
  "match": Object {
    "fieldTest": "valueToMatch",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions rangeQuery 1`] = `
Object {
  "range": Object {
    "fieldTest": Object {
      "lt": 500,
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions termQuery 1`] = `
Object {
  "term": Object {
    "fieldTest.keyword": "valueToMatch",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality Query helper functions termQuery 2`] = `
Object {
  "term": Object {
    "fieldTest": "valueToMatch",
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyGroupFilters Multiple groups 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
        Object {
          "terms": Object {
            "data.attributes.group.keyword": Array [
              "aafc",
              "cnc",
              "seqdb",
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyGroupFilters No groups 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyGroupFilters Single group 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
        Object {
          "term": Object {
            "data.attributes.group.keyword": "aafc",
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyPagination Pagination is correctly applied to query 1`] = `
Object {
  "from": 0,
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "size": 25,
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyPagination Pagination is correctly applied to query 2`] = `
Object {
  "from": 100,
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "size": 100,
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyRootQuery Boolean logic exists, also contains should logic, include the must match minimum. 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "minimum_should_match": 1,
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
      "should": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyRootQuery Boolean logic exists, move it properly to it's own boolean section. 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applyRootQuery No boolean logic exists, just return the query as is. 1`] = `
Object {
  "query": Object {},
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Attempting to sort on a column that does not exist 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {},
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Basic sorting is correctly applied to the query 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn1.keyword": Object {
      "order": "desc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Basic sorting is correctly applied to the query 2`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn1.keyword": Object {
      "order": "asc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Basic sorting is correctly applied to the query 3`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn1.keyword": Object {
      "order": "asc",
    },
    "data.attributes.testColumn2": Object {
      "order": "desc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules No sorting to be performed, leave query the same without adding anything. 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Relationship sorting query generation 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn4": Object {
      "nested_filter": Object {
        "term": Object {
          "included.type": "relationshipType1",
        },
      },
      "nested_path": "included",
      "order": "desc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Relationship sorting query generation 2`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn4": Object {
      "nested_filter": Object {
        "term": Object {
          "included.type": "relationshipType1",
        },
      },
      "nested_path": "included",
      "order": "desc",
    },
    "data.attributes.testColumn5.keyword": Object {
      "nested_filter": Object {
        "term": Object {
          "included.type": "relationshipType2",
        },
      },
      "nested_path": "included",
      "order": "asc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Relationship sorting query generation 3`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "data.attributes.testColumn1.keyword": Object {
      "order": "asc",
    },
    "data.attributes.testColumn2": Object {
      "order": "desc",
    },
    "data.attributes.testColumn4": Object {
      "nested_filter": Object {
        "term": Object {
          "included.type": "relationshipType1",
        },
      },
      "nested_path": "included",
      "order": "desc",
    },
    "data.attributes.testColumn5.keyword": Object {
      "nested_filter": Object {
        "term": Object {
          "included.type": "relationshipType2",
        },
      },
      "nested_path": "included",
      "order": "asc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySortingRules Strings as column definitions 1`] = `
Object {
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
  "sort": Object {
    "testColumn3": Object {
      "order": "asc",
    },
  },
}
`;

exports[`QueryBuilderElasticSearchExport functionality applySourceFiltering Ensure all source filtering are being applied to the query 1`] = `
Object {
  "_source": Array [
    "data.id",
    "data.type",
    "data.attributes.testColumn1",
    "data.attributes.includeThis1",
    "data.attributes.includeThis2",
    "data.attributes.testColumn2",
    "data.attributes.includeThis3",
    "data.attributes.testColumn4",
    "data.attributes.testColumn5",
    "included.id",
    "included.type",
  ],
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "term": Object {
            "data.attributes.materialSampleName.keyword": "Test",
          },
        },
        Object {
          "term": Object {
            "data.attributes.preparationDate": "2022-11-30",
          },
        },
      ],
    },
  },
}
`;
